src/firefly/java/edu/caltech/ipac/util/cache/CachePeerProviderFactory.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
+ */
+
+package edu.caltech.ipac.util.cache;
+
+import edu.caltech.ipac.firefly.messaging.Message;
+import edu.caltech.ipac.firefly.messaging.Messenger;
+import edu.caltech.ipac.firefly.server.util.Logger;
+import edu.caltech.ipac.util.StringUtils;
+import net.sf.ehcache.CacheException;
+import net.sf.ehcache.CacheManager;
+import net.sf.ehcache.Ehcache;
+import net.sf.ehcache.distribution.CacheManagerPeerListener;
+import net.sf.ehcache.distribution.CacheManagerPeerProvider;
+import net.sf.ehcache.distribution.CacheManagerPeerProviderFactory;
+import net.sf.ehcache.distribution.CachePeer;
+import net.sf.ehcache.distribution.RMICacheManagerPeerProvider;
+import net.sf.ehcache.distribution.RMICacheManagerPeerProviderFactory;
+import net.sf.ehcache.util.PropertyUtil;
+
+import java.io.IOException;
+import java.rmi.NotBoundException;
+import java.rmi.RemoteException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+/**
+ * Custom factory for deciding the discovery method to use when creating cache replication.
+ * Currently, peerDiscovery may be "PubSub" or "MultiCast".  When not given, "MultiCast" is used.
+ * Similar to MultiCast, multicastGroupPort is used in PubSub to group related caches for replication.
+ *
+ * Date: 2019-04-08
+ *
+ * @author loi
+ * @version $Id: $
+ */
+public class CachePeerProviderFactory extends CacheManagerPeerProviderFactory {
+    static final Logger.LoggerImpl LOG = Logger.getLogger();
```

Comment:
I am impressed. This took some study to implement. I don't know how many people has done it outside of the ehcache people.